### PR TITLE
Dockerfile: bump alpine from 3.17 to 3.19 and Python 3.10 to 3.11

### PR DESCRIPTION
**Pull request by yuvipanda**
Brings us to latest alpine, and to a newer python

### CODE REVIEW

**Comment by manics** [comment]
What are the benefits of using a virtualenv inside the image?

**Comment by yuvipanda** [comment]
@manics it is mandatory now, it errors out if you don't use one (https://peps.python.org/pep-0668/).

You can see the full config in https://github.com/jupyterhub/repo2docker/actions/runs/7768480733/job/21186449405

```
 #15 [stage-1 3/9] RUN pip3 install hg-evolve --user --no-cache-dir
#15 1.427 error: externally-managed-environment
#15 1.427 
#15 1.427 × This environment is externally managed
#15 1.427 ╰─> 
#15 1.427     The system-wide python installation should be maintained using the system
#15 1.427     package manager (apk) only.
#15 1.427     
#15 1.427     If the package in question is not packaged already (and hence installable via
#15 1.427     "apk add py3-somepackage"), please consider installing it inside a virtual
#15 1.427     environment, e.g.:
#15 1.427     
#15 1.427     python3 -m venv /path/to/venv
#15 1.427     . /path/to/venv/bin/activate
#15 1.427     pip install mypackage
#15 1.427     
#15 1.427     To exit the virtual environment, run:
#15 1.427     
#15 1.427     deactivate
#15 1.427     
#15 1.427     The virtual environment is not deleted, and can be re-entered by re-sourcing
#15 1.427     the activate file.
#15 1.427     
#15 1.427     To automatically manage virtual environments, consider using pipx (from the
#15 1.427     pipx package).
#15 1.427 
#15 1.427 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
#15 1.427 hint: See PEP 668 for the detailed specification.
#15 ERROR: process "/bin/sh -c pip3 install hg-evolve --user --no-cache-dir" did not complete successfully: exit code: 1
```

File: Dockerfile [hunk 1]
```diff
@@ -1,30 +1,40 @@
 # syntax = docker/dockerfile:1.3
-ARG ALPINE_VERSION=3.17
+ARG ALPINE_VERSION=3.19
 FROM alpine:${ALPINE_VERSION}
 
 RUN apk add --no-cache git python3 python3-dev py3-pip py3-setuptools build-base
 
 # build wheels in first image
```
Comment (consideRatio) [inline Dockerfile:7]:
```suggestion
# build wheels in a build stage
```

File: Dockerfile [hunk 2]
```diff
@@ -1,30 +1,40 @@
...
+ && pip wheel --no-cache-dir /tmp/src \
  && ls -l /tmp/wheelhouse
 
 FROM alpine:${ALPINE_VERSION}
 
 # install python, git, bash, mercurial
 RUN apk add --no-cache git git-lfs python3 py3-pip py3-setuptools bash docker mercurial
 
+ENV VIRTUAL_ENV /opt/venv
+ENV PATH ${VIRTUAL_ENV}/bin:${PATH}
```
Comment (consideRatio) [inline Dockerfile:27]:
```suggestion
ENV VIRTUAL_ENV=/opt/venv
ENV PATH=${VIRTUAL_ENV}/bin:${PATH}
```

Outdated syntax, may be removed, see https://docs.docker.com/reference/dockerfile/#env.

File: Dockerfile [hunk 3]
```diff
@@ -1,30 +1,40 @@
...
-ARG ALPINE_VERSION=3.17
+ARG ALPINE_VERSION=3.19
 FROM alpine:${ALPINE_VERSION}
 
 RUN apk add --no-cache git python3 python3-dev py3-pip py3-setuptools build-base
 
-# build wheels in first image
+# build wheels in a build stage
+ENV VIRTUAL_ENV /opt/venv
+ENV PATH ${VIRTUAL_ENV}/bin:${PATH}
```
Comment (consideRatio) [inline Dockerfile:9]:
```suggestion
ENV VIRTUAL_ENV=/opt/venv
ENV PATH=${VIRTUAL_ENV}/bin:${PATH}
```

File: Dockerfile [hunk 4]
```diff
@@ -1,30 +1,40 @@
 # syntax = docker/dockerfile:1.3
-ARG ALPINE_VERSION=3.17
+ARG ALPINE_VERSION=3.19
 FROM alpine:${ALPINE_VERSION}
 
 RUN apk add --no-cache git python3 python3-dev py3-pip py3-setuptools build-base
 
-# build wheels in first image
+# build wheels in a build stage
+ENV VIRTUAL_ENV=/opt/venv
```
Comment (consideRatio) [inline Dockerfile:8]:
```suggestion
ARG VIRTUAL_ENV=/opt/venv
```

File: Dockerfile [hunk 5]
```diff
@@ -1,30 +1,40 @@
...
+ && pip install wheel \
+ && pip wheel --no-cache-dir /tmp/src \
  && ls -l /tmp/wheelhouse
 
 FROM alpine:${ALPINE_VERSION}
 
 # install python, git, bash, mercurial
 RUN apk add --no-cache git git-lfs python3 py3-pip py3-setuptools bash docker mercurial
 
+ENV VIRTUAL_ENV=/opt/venv
```
Comment (consideRatio) [inline Dockerfile:26]:
```suggestion
ARG VIRTUAL_ENV=/opt/venv
```

**Comment by consideRatio** [comment]
Merging this resolves repo2docker-action https://github.com/jupyterhub/repo2docker-action/issues/113.

I pushed commits to:
- tweak comment unrelated to PR
- use modern ENV syntax with `=` signs between key/values
- use ARG instead of ENV when not needed to provide an env var only during docker build as compared to also during docker run

**Review by consideRatio — APPROVED** [review]
If tests pass, this LGTM.

**Comment by consideRatio** [comment]
I marked this as breaking for possible highlight its a possible breaking change for users of the image

**Pull request merged**